### PR TITLE
[MPS] Fix tensor with non-zero storage offset graph gathering

### LINKED ISSUE #87856: torch.stack gives wrong results on MPS
### 🐛 Describe the bug

`torch.stack` yields different results on `cpu` vs `mps` for certain input shapes/stack dimensions: 

```python

import torch

x = torch.tensor([[1,2]])
for device in ('cpu', 'mps'):
    y = x.to(device)
    z = torch.stack((y[:,:1], y[:,-1:]), dim=-1)
    print(z)
```

```
tensor([[[1, 2]]])
tensor([[[1, 1]]], device='mps:0')
```

Above is on current nightly but it seemed that the same bug persists for older versions too.

### Versions

```
PyTorch version: 1.14.0.dev20221027
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.3.1 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2)
CMake version: version 3.21.2
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:01:00)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-12.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.4
[pip3] torch==1.14.0.dev20221027
[conda] numpy                     1.23.4           py39hefdcf20_1    conda-forge
[conda] torch                     1.14.0.dev20221027          pypi_0    pypi
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### LINKED ISSUE #91065: [MPS] Incorrect results of two multiplied chunked tensors produced by `unsafe_chunk`
### 🐛 Describe the bug

This causes the inconsistent results mentioned in #90421. `torch.chunk` has no issue.

```python
import torch
a = torch.rand(5, dtype=torch.float32, device="cpu")
ret = a.unsafe_chunk(4, 0)
y = ret[0] * ret[2]
a_mps = a.to("mps")
ret_mps = a_mps.unsafe_chunk(4, 0)
y_mps = ret_mps[0] * ret_mps[2]
torch.testing.assert_close(y, y_mps, check_device=False)
```

### Versions

```
PyTorch version: 2.0.0a0+giteae0f3f
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6.1 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.24.0
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:00:33)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-12.6.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

Versions of relevant libraries:
[pip3] mypy==0.960
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] torch==2.0.0a0+giteae0f3f
[pip3] torchvision==0.13.0
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] torch                     2.0.0a0+giteae0f3f           dev_0    <develop>
[conda] torchvision               0.13.0                   pypi_0    pypi
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by qqaatw**
Previously, the "can slice" flag in Placeholder constructor in `OperationUtils.mm` is conditioned on whether the numbers of dimensions of base shape and view shape are the same. This doesn't consider the situation that a view tensor could be the base tensor's sliced and then unsqueezed version, resulting in different num of dims.

For example, if we want to stack `y_mps` and `x_mps` on the last dim:
```
t_mps = torch.tensor([1, 2, 3, 4], device="mps")
x_mps = t_mps[2:]  # [3, 4]
y_mps = t_mps[:2]  # [1, 2]

res_mps = torch.stack((y_mps, x_mps), dim=-1)
```

the kernel will unsqueeze both of them on the last dim and then concatenate them, which is equivalent to:

```
res_mps = torch.cat((y_mps.unsqueeze(-1), x_mps.unsqueeze(-1)), dim=-1)
```

`x_mps.unsqueeze(-1)` is an unsqueezed and contiguous tensor with a storage offset, this kind of tensors should be sliceable without cloning its storage.

Fixes #87856
Fixes #91065

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### CODE REVIEW

File: test/test_mps.py [hunk 1]
```diff
@@ -1406,25 +1406,28 @@ def helper(operator):
             if operator == "<=":
                 res_mps = x_mps <= y_mps
                 res_cpu = x_cpu <= y_cpu
-            if operator == "<":
+            elif operator == "<":
```
Comment (malfet) [inline test/test_mps.py:1409]: This seems unrelated to change in question, is it?

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 2]
```diff
@@ -226,7 +226,7 @@ void printTensorNDArray(const Tensor& t) {
   // extract the pointer to MTLBuffer from the Tensor's storage
   id<MTLBuffer> srcBuf = getMTLBufferStorage(src);
   // a view tensor could be contiguous (e.g., slice ops) or non-contiguous (e.g., transpose())
-  if (src.is_view() || !src.is_contiguous()) {
+  if (src.is_view() || src.storage_offset() != 0 || !src.is_contiguous()) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/OperationUtils.mm:229]: Can you please add a more detailed description, why is this necessary? Shouldn't we just add handling for non-zero storage_offset to the MTLBuffer pointer computation?

**Review by malfet — CHANGES_REQUESTED** [review]
This PR contains unrelated changes as well as not much of an explanation as to why it is necessary to make a copy of the tensor view. 

I.e. it probably fixes the symptom, but not the fundamental underlying problem.

File: test/test_mps.py (hunk 1 above)
Comment (qqaatw) [reply test/test_mps.py:1409]: Yes not much related, but I think this improvement is trivial and small enough to not cause a reading difficulty. I feel it would be more costly and unnecessary to improve this in a separate PR, wouldn't it?

File: aten/src/ATen/native/mps/OperationUtils.mm (hunk 2 above)
Comment (qqaatw) [reply aten/src/ATen/native/mps/OperationUtils.mm:229]: Let me find if there is a better way to handle this kind of tensor.

Comment (qqaatw) [reply aten/src/ATen/native/mps/OperationUtils.mm:229]: For now I didn't find a proper way to get an offset of a MTLBuffer (from the [docs](https://developer.apple.com/documentation/metal/mtlbuffer?language=objc) on Appls's website there seems no such methods provided, not sure whether using primitive pointer operations to it is safe or not). What I know is that `MTLComputeCommandEncoder` derived from a `MTLCommandBuffer` in a custom metal kernel can add an offset for reading an existing buffer, like so `[computeEncoder setBuffer:a offset:0 atIndex:0];`, but it seems to not apply to this case. So unfortunately we may have to handle this as is.

Please let me know if I'm missing something.

Comment (kulinseth) [reply aten/src/ATen/native/mps/OperationUtils.mm:229]: There is a way to add offset by creating Array view. You will have to tile the offset to all the dimensions as its done in the [code path ](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/mps/operations/View.mm#L488).

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 3]
```diff
@@ -289,7 +289,7 @@ void printTensorNDArray(const Tensor& t) {
   } else {
     if (!mpsShape) {
       mpsShape = getMPSShape(_tensor);
-  }
+    }
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/OperationUtils.mm:292]: Fix this.

**Review by kulinseth — APPROVED** [review]

**Comment by kulinseth** [comment]
Modulo the nit, changes look good.

File: aten/src/ATen/native/mps/OperationUtils.mm (hunk 3 above)
Comment (qqaatw) [reply aten/src/ATen/native/mps/OperationUtils.mm:292]: This is purposely an indentation fix for `if (!mpsShape) { ` block.
